inox2d/src/render.rs

```diff
@@ -218,3 +222,173 @@ impl Puppet {
+                let mask_node_render_ctx = &puppet.render_ctx.node_render_ctxs[&mask.source];
+
+                match (&mask_node.data, &mask_node_render_ctx.kind) {
+                    (
+                        InoxData::Part(ref mask_part),
+                        RenderCtxKind::Part(ref mask_part_render_ctx),
+                    ) => {
+                        self.draw_part_self(
+                            true,
+                            camera,
+                            mask_node_render_ctx,
+                            mask_part,
+                            mask_part_render_ctx,
+                        )?;
+                    }
+
+                    (
+                        InoxData::Composite(ref mask_composite),
+                        RenderCtxKind::Composite(ref mask_children),
+                    ) => {
+                        self.draw_composite(true, camera, mask_composite, puppet, mask_children)?;
+                    }
+
+                    _ => {
+                        // This match block clearly is sign that the data structure needs rework
+                        todo!();
+                    }
+                }
+            }
+            self.on_begin_masked_content()?;
+            self.draw_part_self(false, camera, node_render_ctx, part, part_render_ctx)?;
+            self.on_end_mask()?;
+        } else {
+            self.draw_part_self(false, camera, node_render_ctx, part, part_render_ctx)?;
+        }
+        Ok(())
+    }
+
+    fn begin_composite(&self) -> Result<(), Self::Error>;
+    fn end_composite(&self) -> Result<(), Self::Error>;
+
+    fn draw_composite_self(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error> {
+        self.begin_composite()?;
+
+        for &uuid in children {
+            let node = puppet.nodes.get_node(uuid).unwrap();
+            let node_render_ctx = &puppet.render_ctx.node_render_ctxs[&uuid];
+
+            if let (InoxData::Part(ref part), RenderCtxKind::Part(ref part_render_ctx)) =
+                (&node.data, &node_render_ctx.kind)
+            {
+                if as_mask {
+                    self.draw_part_self(true, camera, node_render_ctx, part, part_render_ctx)?;
+                } else {
+                    self.draw_part(camera, node_render_ctx, part, part_render_ctx, puppet)?;
+                }
+            } else {
+                // composite inside composite simply cannot happen
+            }
+        }
+
+        self.end_composite()?;
+        Ok(())
+    }
+
+    fn draw_composite(
+        &self,
+        as_mask: bool,
+        camera: &Mat4,
+        composite: &Composite,
+        puppet: &Puppet,
+        children: &[InoxNodeUuid],
+    ) -> Result<(), Self::Error>;
+
+    fn draw_drawables(&self, camera: &Mat4, puppet: &Puppet) -> Result<(), Self::Error> {
```

Comment:
Naming of course can be changed. The only comment I want to make is that this call iterates over all the "top-level" drawables, which is defined at [source/inochi2d/core/puppet.d:212, branch v0_8](https://github.com/Inochi2D/inochi2d/blob/273dd40f2e406e9044ccfde2bf77999b15369b3b/source/inochi2d/core/puppet.d#L212C1-L218C22) as

```d
/**
    A list of parts that are not masked by other parts

    for Z sorting
*/
@Ignore
Node[] rootParts;
```

Consequentially speaking this call renders the whole puppet, so yeah maybe we can eventually use a simple name like `draw()`, and leave the drawable remarks to the comment.

Reply:
Yeah. That it iterates over all the root parts and z-sorted is an implementation detail, it's better to be explicit about what it does in the end.

Reply:
I think we can one day ask Luna about this. From the actual implementations of the ref impl `rootParts` seems like just simply the `top-level` `Part`s and `Composite`s, excluding `Mask`s and `Composite` children. This set has nothing to do with "not masked by others". This makes sense to me, and the current inox impl is doing this...

...almost. I now find a bug, where composite children are present in the `nodes_zsorted`, thus `drawables_zsorted`. The render seems to work now because Aka chan is not that complex, but this is definitely wrong: A composite children will be drawn during the composite draw, and will be drawn again independently as a normal `Part`.